File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCaseCompletable.java
package com.dimatest.movieapp.common;

import androidx.core.util.Preconditions;
import io.reactivex.Completable;
import io.reactivex.Scheduler;
import io.reactivex.observers.DisposableCompletableObserver;

public abstract class BaseUseCaseCompletable<Params> extends BaseUseCase{

    public BaseUseCaseCompletable(Scheduler mainThread, Scheduler ioThread) {
        super(mainThread, ioThread);
    }

    abstract protected Completable buildUseCaseObservable(Params params);

    public void execute(Params params, DisposableCompletableObserver observer) {
        Preconditions.checkNotNull(observer);
        final Completable observable = this.buildUseCaseObservable(params)
                .subscribeOn(ioThread)
                .observeOn(mainThread)
                .doOnSubscribe(__ -> isLoading.postValue(true))
                .doOnTerminate(() -> isLoading.postValue(false))
                .doOnDispose(() -> isLoading.postValue(false));
        addDisposable(observable.subscribeWith(observer));
    }
}

File: app/src/main/java/com/dimatest/movieapp/common/BaseUseCaseCompletable.kt
package com.dimatest.movieapp.common

import io.reactivex.Completable
import io.reactivex.Scheduler
import io.reactivex.observers.DisposableCompletableObserver

abstract class BaseUseCaseCompletable<Params>(
        mainThread: Scheduler, ioThread: Scheduler
) : BaseUseCase(mainThread, ioThread) {

    protected abstract fun buildUseCaseObservable(params: Params): Completable

    fun execute(params: Params, observer: DisposableCompletableObserver) {
        val observable = buildUseCaseObservable(params)
                .subscribeOn(ioThread)
                .observeOn(mainThread)
                .doOnSubscribe { isLoading.postValue(true) }
                .doOnTerminate { isLoading.postValue(false) }
                .doOnDispose { isLoading.postValue(false) }
        addDisposable(observable.subscribeWith(observer))
    }
}